app/lti_controller.rb

```diff
@@ -128,4 +129,34 @@ def tool_proxy_signed_post(tool_proxy, endpoint)
       }
     }
   end
+
+
+  ##
+  #  Used to determine if the nonce is still valid
+  #
+  #  +nonce+:: This is the cache key used to check if the nonce key has been used
+  #  +timestamp+:: The timestamp of when the request was signed
+  #  +nonce_age+:: An ActiveSupport::Duration describing how old a nonce can be
+  #
+  #  The +nonce_age+ creates a range that the timestamp must fall between for the nonce to be valid
+  #  valid_range = +Time.now+ - (the +nonce_age+ duration)
+  #  i.e. if the current time was 2010-04-23T12:30:00Z and the +nonce_age+ was 30min
+  #  then the valid time range that the timestamp must fall between would
+  #  be "2010-04-23T12:30:00Z/2010-04-23T13:00:00Z"
+  #
+  #  =Time line Examples for valid and invalid timestamps
+  #
+  #  |---nonce_age---timestamp---Time.now---|  VALID
+  #
+  #  |---timestamp---nonce_age---Time.now---| INVALID
+  #
+  #  |---nonce_age---Time.now---timestamp---| INVALID
+  #
+  def check_and_store_nonce(nonce, timestamp, nonce_age)
+    allowed_future_skew = 60.seconds
+    valid = timestamp.between?(nonce_age.ago.to_i, (Time.now + allowed_future_skew).to_i)
```

Comment:
this line could be extracted to outside the method, but I'm hesitant to separate it since it is tightly related to nonce checking